src/ray/rpc/test/grpc_server_client_test.cc

```diff
@@ -0,0 +1,233 @@
+  request.set_sleep_time_ms(1);
+  bool call_timed_out = false;
+  Sleep(request, [&call_timed_out](const Status status, const SleepReply &reply) {
+    RAY_LOG(INFO) << "Replied, status=" << status;
+    ASSERT_TRUE(status.IsIOError());
+    call_timed_out = true;
+  });
+  // Wait for clinet call timed out.
+  while (!call_timed_out) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+  // Unfreeze server so the server thread can exit.
+  test_service_handler_.frozen = false;
+  while (test_service_handler_.reply_failure_count <= 0) {
+    RAY_LOG(INFO) << "Waiting for reply failure";
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+}
+
+// This test aims to test ServerCall leaking when client died before server's reply.
+// Check https://github.com/ray-project/ray/pull/17863 for more context
+TEST_F(TestGrpcServerClientFixture, TestClientDiedBeforeReply) {
+  // Reinit ClientCallManager with short timeout, so that call won't block.
+  grpc_client_.reset();
+  client_call_manager_.reset();
+  client_call_manager_.reset(new ClientCallManager(client_io_service_, /*num_thread=*/1,
+                                                   /*call_timeout_ms=*/100));
+  grpc_client_.reset(
+      new GrpcClient<TestService>("localhost", 123321, *client_call_manager_));
+  // Freeze server first, it won't reply any request.
+  test_service_handler_.frozen = true;
+  // Send request.
+  SleepRequest request;
+  request.set_sleep_time_ms(1);
+  bool call_timed_out = false;
+  Sleep(request, [&call_timed_out](const Status status, const SleepReply &reply) {
+    RAY_LOG(INFO) << "Replied, status=" << status;
+    ASSERT_TRUE(status.IsIOError());
+    call_timed_out = true;
+  });
+  // Wait for clinet call timed out. Client socket won't be closed until all calls are
+  // timed out.
+  while (!call_timed_out) {
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+  // Shutdown client before server replies.
+  grpc_client_.reset();
+  client_call_manager_.reset();
+  // Unfreeze server, server will fail to reply.
+  test_service_handler_.frozen = false;
+  while (test_service_handler_.reply_failure_count <= 0) {
+    RAY_LOG(INFO) << "Waiting for reply failure";
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+  // Reinit client with infinite timeout.
+  client_call_manager_.reset(new ClientCallManager(client_io_service_));
+  grpc_client_.reset(
+      new GrpcClient<TestService>("localhost", 123321, *client_call_manager_));
+  // Send again, this request should be replied. If any leaking happened, this call won't
+  // be replied to since the max_active_rpcs is 1.
+  bool done = false;
+  Sleep(request, [&done](const Status status, const SleepReply &reply) {
+    RAY_LOG(INFO) << "replied, status=" << status;
+    done = true;
+  });
+  while (!done) {
+    RAY_LOG(INFO) << "waiting";
+    std::this_thread::sleep_for(std::chrono::milliseconds(1000));
+  }
+}
```

Comment:
One last request here; can we add a test for backpressure?

Reply:
Maybe it's better to add this test in a separate PR? Because it's not related to this issue. I don't want to make this PR too large.

I'll open another PR to do this, do you think it's OK?